docs/architecture.md

```diff
@@ -0,0 +1,176 @@
+
+To do this, GraphSync maintains several independent threads of execution (i.e. goroutines). Specifically:
+- On the requestor side:
+1. We maintain an independent thread to make and track requests (RequestManager)
+2. We maintain an independent thread to feed incoming blocks to selector verifications (AsyncLoader)
+3. Each outgoing request has an independent thread performing selector verification
+4. Each outgoing request has an independent thread collecting and buffering final responses before they are returned to the caller. Graphsync returns responses to the caller through a channel. If the caller fails to immediately read the response channel, this should not block other requests from being processed.
+- On the responder side:
+1. We maintain an independent thread to receive incoming requests and track outgoing responses. As each incoming request is received, it's put into a prioritized queue.
+2. We maintain fixed number of threads that continuosly pull the highest priority request from the queue and perform the selector query for that request
+3. Each peer we respond to has an independent thread thread marshalling and deduplicating outgoing responses and blocks before they are sent back. This minimizes data sent on the wire and allows queries to proceed without getting blocked by the network.
+- At the messaging layer:
+1. Each peer we send messages to has an independent thread collecting and buffering message data while waiting for the last message to finish sending. This allows higher level operations to execute without getting blocked by a slow network
+
+The following diagram illustrates concurrent threads operating as a client makes calls to GraphSync and messages arive from the network:
+![GraphSync Process Diagram](./processes.png)
+
+The remaining sections of this document outline internal workings of major graphsync components in more detail. 
+
+## Dependencies
+
+### Network Implementation
+
+The network implementation needs to provide basic lower level utilties for sending and receiving messages. A default implementation using `libp2p` is included in the package, and a mock version is provided for testing. 
+
+### Bridge To IPLD
+
+Rather than interact with `go-ipld-prime` directly, `go-graphsync` makes calls via a bridge interface provided as a dependency. During the initial development of `go-graphsync`, key components of `go-ipld-prime` were not finished or changing rapidly. The bridge provides a way to keep the interfaces somewhat stable from `go-graphsyncs` standpoint, and a mechanism to provide a mock version of go-ipld-prime for testing. This allowed `go-graphsync` to be written at the same time as `go-ipld-prime`. As `go-ipld-prime` stabilizes, it make make sense to remove this interface, though it is still useful for test isolation. The library provides a default bridge as well as a mock bridge.
```

Comment:
```suggestion
Rather than interact with `go-ipld-prime` directly, `go-graphsync` makes calls via a bridge interface provided as a dependency. During the initial development of `go-graphsync`, key components of `go-ipld-prime` were not finished or changing rapidly. The bridge provides a way to keep the interfaces somewhat stable from `go-graphsyncs` standpoint, and a mechanism to provide a mock version of go-ipld-prime for testing. This allowed `go-graphsync` to be written at the same time as `go-ipld-prime`. As `go-ipld-prime` stabilizes, it might make sense to remove this interface, though it is still useful for test isolation. The library provides a default bridge as well as a mock bridge.
```